### ISSUE TITLE
[Discussion] Serving CUJ and examples

### ISSUE THREAD

**Issue by lluunn** [kubeflow/kubeflow#1004]
We should define a serving CUJ, and have an example for it.

For CUJ, how about the following:
- User develops a model (in notebook, using Kubeflow tfjob, or from somewhere else) and exports it to GCS or S3 or some other FS.
- User creates a ksonnet app to describe deployment of the model (can specify GPU, number of cores etc)
- User deploys the model, and starts sending prediction requests (HTTP or gRPC)
- User monitors the model (metrics exported and available, for example, number of errors/requests, latencies)
- User develops another model, and rolls out the new model. X% of traffic is directed to the new model. User compares performance. If new model is better, it's rolled out completely. If new model is worse, it's rolled back.

**Comment by johnugeorge** [comment]
We can have  a higher level resource that manages the route and model version.  This resource can specify deployed model versions and their corresponding route %.  Versions may also be generated automatically.  By this, user needn't worry about model versioning and routes are automatically updated when a model version is deployed.

**Comment by amsaha** [comment]
Is this similar to this: https://github.com/SeldonIO/seldon-core ? Their recent KubeCon presentation on this: https://www.youtube.com/watch?v=pDlapGtecbY

**Comment by cliveseldon** [comment]
I work at Seldon and thus on seldon-core. 

We are presently building up some examples for CI/CD integration and istio user journeys such as canary rollouts. Our ML Service graphs certainly allow rolling updates of new versions of models already and fit nicely into "gitops" with these being declarative representations of your ML serving architecture you can commit to source control.

Would be happy to discuss how we can build some user journeys in kubeflow that show this integrated into training.

**Comment by jlewi** [comment]
I'd like to figure out what we want the state of serving to be at the end of 0.3.0.
Seldon/TFServing/etc... they are all important pieces of that.

Here's our current [backlog](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aopen+is%3Aissue+org%3Akubeflow+label%3Arelease%2F0.3.0++label%3Aarea%2Finference)

Are there other issues we should add? Should we cross reference issues in Seldon?

**Comment by cfregly** [comment]
this functionality is also available in [PipelineAI](https://pipeline.ai) here:  http://quickstart.pipeline.ai/kubernetes 

we're using istio for the routing, traffic mirroring/shadowing, and Netflix-style Chaos Engineering (we come from Netflix).

**Comment by cliveseldon** [comment]
@jlewi Yes if you could cross reference serving issues on to [seldon](https://github.com/SeldonIO/seldon-core/issues) that would be very useful for us to ensure we cover kubeflow's 0.3 goals for serving.

**Comment by jlewi** [comment]
@cliveseldon I went through the seldon backlog and cross referenced a couple issues.

@cfregly I took a look at PipelineAI but I only saw one open issue.

It would be great to find a way to build a shared vision and work backwards from that.

An intern @activatedgeek is working with @hamelsmu to built out a more complicated E2E example:
http://bit.ly/gh-kf-search

In particular for this example, we'd like to highlight

   1. Collecting feedback about predictions (e.g. does user click on them)
   1. Rolling out new models into prediction.

It would be nice to align on these areas e.g.

- GitOps for non model deployment
   - kubeflow/kubeflow#1118
   - SeldonIO/seldon-core#67 
   - SeldonIO/seldon-core#11

- Using Seldon to create a model server for nmslib

One question I have is whether we could use Seldon for model deployment with other model servers; e.g. TFServing, [NVIDIA DL Server](https://github.com/NVIDIA/dl-inference-server) or Onyx?
